Resolve remaining parts of CSS gradient specifiers to Lengths at style builder time

weinig opened the pull request:
#### a33bb8bc70a1b75e1e45a0a85fb299e62a45005c
<pre>
Resolve remaining parts of CSS gradient specifiers to Lengths at style builder time
<a href="https://bugs.webkit.org/show_bug.cgi?id=275667">https://bugs.webkit.org/show_bug.cgi?id=275667</a>

Reviewed by Darin Adler.

Updates style gradient representations to stop using CSSValues and instead
have Lengths (or other raw values) computed at style builder time.

Additionally, a few uses of CSSValues in the CSS gradient representation
were replaced with more direct values (Raw/UnevaluatedCalc) where we have
good support (that is, all the non-&lt;length-percentage&gt; places). Once there
is additional work for supporting &lt;length-percentage&gt;, further strong typing
can be added.

* LayoutTests/fast/gradients/css3-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/css3-gradient-parsing.html:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing.html:
    - Update tests and results to account for proper computed value serialization of positions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
    - Update test to latest in WPT, which we now pass again.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-linear-radial-conic-gradient-001-expected.txt:
    - Update results for now passing test.

* Source/WTF/wtf/MathExtras.h:
    - Add additional radian conversions and remove unnecessary &quot;inline&quot; specifiers.

* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::CSSPrimitiveValue::computeRadians):
    - Add helper akin to computeDegrees to directly compute radians.

* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
    - Update to use more strongly typed representations and to process
      values into lengths at style builder time.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
(WebCore::evaluateCalc):
    - Move generic catchall above other uses so it can always be found.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
    - Update consumers to produce new CSS representation.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
    - Update specifiers to have computed Length values, removing
      most reliance on CSS gradient types (except enums).

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
    - Update to use Lengths.

Canonical link: <a href="https://commits.webkit.org/280278@main">https://commits.webkit.org/280278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35ebcbf7f7542ec9f5bbfa7ad1d228aa84fc4ae0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56138 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35464 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6574 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6768 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45277 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4556 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33352 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48422 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26321 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/55664 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30133 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5753 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5578 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49216 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52107 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6024 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61427 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55375 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/46 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6149 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52674 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/46 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48490 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52395 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12438 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/input-targets-root-while-render-blocked.html (failure)") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/41 "Passed tests") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/77135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8332 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31291 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/77135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32377 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33460 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32124 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
```diff
@@ -23,13 +23,13 @@
-    { input: 'ellipse 50% 40em', output: '50% 40em' },
-    { input: 'at right center' },
+    { input: 'ellipse 50% 40em', output: '50% 640px' },
+    { input: 'at right center', output: 'at 100% 50%' },
 ];
 
 const CONIC_GRADIENT_SPECIFIERS = [
     { input: 'from 30deg' },
-    { input: 'at left 10px top 50em' },
+    { input: 'at left 10px top 50em', output: 'at 10px 800px' },
```
Does this need to be upstreamed?

weinig replied:
No, these were from upstream.

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -124,22 +125,26 @@ constexpr float turnsPerDegreeFloat = 1.0f / 360.0f;
+constexpr float deg2turn(float d) { return d * turnsPerDegreeFloat; }
+constexpr float turn2deg(float t) { return t * degreesPerTurnFloat; }
+
+// Treat theses as conversions through the canonical unit for angles, which is degrees.
```
theses -> these

darinadler commented on Source/WebCore/css/CSSGradientValue.cpp:
```diff
@@ -38,17 +39,45 @@
+
+static bool styleImageIsUncacheable(const std::optional<CSSGradientPosition>& position)
+{
+    return position ? styleImageIsUncacheable(*position) : false;
```
I would use `&&` instead of `? :` here.

darinadler commented on Source/WebCore/css/CSSGradientValue.cpp:
```diff
@@ -453,23 +681,26 @@ bool CSSPrefixedLinearGradientValue::equals(const CSSPrefixedLinearGradientValue
+    serializationForCSS(result, m_data.first);
+    result.append(", "_s);
+    serializationForCSS(result, m_data.second);
+
```
I don’t think the blank lines are helpful.

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -1124,55 +1277,50 @@ Ref<Gradient> StyleGradientImage::createGradient(const DeprecatedLinearData& lin
 // MARK: - Radial create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const RadialData& radial, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const RadialData& radial, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -1102,18 +1261,12 @@ Ref<Gradient> StyleGradientImage::createGradient(const PrefixedLinearData& linea
 // MARK: - Deprecated Linear create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const DeprecatedLinearData& linear, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const DeprecatedLinearData& linear, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -1034,57 +1199,51 @@ Ref<Gradient> StyleGradientImage::createGradient(const LinearData& linear, const
 // MARK: - Prefixed Linear create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const PrefixedLinearData& linear, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const PrefixedLinearData& linear, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -977,47 +1148,41 @@ static inline float horizontalEllipseRadius(const FloatSize& p, float aspectRati
 // MARK: - Linear create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const LinearData& linear, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const LinearData& linear, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -1279,51 +1427,45 @@ Ref<Gradient> StyleGradientImage::createGradient(const RadialData& radial, const
 // MARK: - Prefixed Radial create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const PrefixedRadialData& radial, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const PrefixedRadialData& radial, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -1397,21 +1547,15 @@ Ref<Gradient> StyleGradientImage::createGradient(const PrefixedRadialData& radia
 // MARK: - Deprecated Radial create.
 
-Ref<Gradient> StyleGradientImage::createGradient(const DeprecatedRadialData& radial, const RenderElement& renderer, const FloatSize& size, const RenderStyle& style) const
+Ref<Gradient> StyleGradientImage::createGradient(const DeprecatedRadialData& radial, const RenderElement&, const FloatSize& size, const RenderStyle& style) const
```
Can we eliminate the `renderer` argument entirely instead of passing it and ignoring it?

weinig committed:
Resolve remaining parts of CSS gradient specifiers to Lengths at style builder time
https://bugs.webkit.org/show_bug.cgi?id=275667

Reviewed by Darin Adler.

Updates style gradient representations to stop using CSSValues and instead
have Lengths (or other raw values) computed at style builder time.

Additionally, a few uses of CSSValues in the CSS gradient representation
were replaced with more direct values (Raw/UnevaluatedCalc) where we have
good support (that is, all the non-<length-percentage> places). Once there
is additional work for supporting <length-percentage>, further strong typing
can be added.

* LayoutTests/fast/gradients/css3-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/css3-gradient-parsing.html:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing.html:
    - Update tests and results to account for proper computed value serialization of positions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
    - Update test to latest in WPT, which we now pass again.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-linear-radial-conic-gradient-001-expected.txt:
    - Update results for now passing test.

* Source/WTF/wtf/MathExtras.h:
    - Add additional radian conversions and remove unnecessary "inline" specifiers.

* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::CSSPrimitiveValue::computeRadians):
    - Add helper akin to computeDegrees to directly compute radians.

* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
    - Update to use more strongly typed representations and to process
      values into lengths at style builder time.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
(WebCore::evaluateCalc):
    - Move generic catchall above other uses so it can always be found.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
    - Update consumers to produce new CSS representation.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
    - Update specifiers to have computed Length values, removing
      most reliance on CSS gradient types (except enums).

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
    - Update to use Lengths.

Canonical link: https://commits.webkit.org/280278@main
